pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -1138,17 +1138,28 @@ func (c *Controller) servicesForNamespacedName(name types.NamespacedName) []*mod
 	return nil
 }
 
-func serviceUpdateNeedsPush(prev, curr *model.Service) bool {
+func serviceUpdateNeedsPush(prev, curr *v1.Service, preConv, currConv *model.Service) bool {
 	if !features.EnableOptimizedServicePush {
 		return true
 	}
-	if prev == nil {
-		return !curr.Attributes.ExportTo.Contains(visibility.None)
+	if preConv == nil {
+		return !currConv.Attributes.ExportTo.Contains(visibility.None)
 	}
 	// if service are not exported, no need to push
-	if prev.Attributes.ExportTo.Contains(visibility.None) &&
-		curr.Attributes.ExportTo.Contains(visibility.None) {
+	if preConv.Attributes.ExportTo.Contains(visibility.None) &&
+		currConv.Attributes.ExportTo.Contains(visibility.None) {
 		return false
 	}
-	return !prev.Equals(curr)
+	if !preConv.Equals(currConv) {
+		return true
+	}
+	if prev != nil && curr != nil {
+		// check if target ports are changed
+		for i := 0; i < len(prev.Spec.Ports); i++ {
+			if prev.Spec.Ports[i].TargetPort != curr.Spec.Ports[i].TargetPort {
```

Comment:
could this panic if ports was removed

Reply:
+1. 

Also do we want to compare just targetPort? or the full spec? I guess `conv.Equals` check handles anything else -- but maybe we can comment that in the code?

Reply:
> could this panic if ports was removed

`preConv.Equals(currConv)` already makes sure the length of ports is not changed